compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala

```diff
@@ -2469,44 +2469,57 @@ abstract class GenJSCode extends plugins.PluginComponent
 
         // Binary operation
         case List(lsrc_in, rsrc_in) =>
-          def convertArg(tree: js.Tree, tpe: Type) = {
-            val kind = toTypeKind(tpe)
+          val leftKind = toTypeKind(args(0).tpe)
+          val rightKind = toTypeKind(args(1).tpe)
+
+          val opType = (leftKind, rightKind) match {
+            case (DoubleKind, _) | (_, DoubleKind) => jstpe.DoubleType
+            case (FloatKind, _) | (_, FloatKind)   => jstpe.FloatType
+            case (INT(_), _) | (_, INT(_))         => jstpe.IntType
+            case (BooleanKind, BooleanKind)        => jstpe.BooleanType
+            case _                                 => jstpe.AnyType
+          }
 
-            // If we end up with a long, target must be float or double
-            val fromLong =
+          def convertArg(tree: js.Tree, kind: TypeKind) = {
+            /* If we end up with a long, the op type must be float or double,
+             * so we can first eliminate the Long case by converting to Double.
+             */
+            val notLong =
               if (kind == LongKind) js.UnaryOp(js.UnaryOp.LongToDouble, tree)
               else tree
 
-            if (resultType != jstpe.FloatType) fromLong
-            else if (kind == FloatKind) fromLong
-            else js.UnaryOp(js.UnaryOp.DoubleToFloat, fromLong)
+            if (opType != jstpe.FloatType) notLong
```

Comment:
Could we make this `opType == jstpe.DoubleType` instead? Seems easier to understand.